### PR TITLE
Fix update of viscous strains in linear viscoelastic materials 

### LINKED ISSUE #11064: Update of internal viscous strains for linear viscoelastic materials use the wrong tensor
## Rationale

In ``GeneralizedKelvinVoingBase::updateQpViscousStrains`` the current values of the elasticity tensors of the different springs are used to update the internal viscoelastic strains. However, this update corresponds to the increase in creep strains from the previous time step and therefore should use the previous value of the elasticity tensors instead.

For the same reason, the previous value of the driving eigenstrain needs to be used instead of the current value in both ``GeneralizedKelvinVoigtBase::updateQpViscousStrains`` and ``GeneralizedMaxwellBase::updateQpViscousStrains``.

This mistake does not appear with the current viscoelastic models ``GeneralizedKelvinVoigtModel`` and ``GeneralizedMaxwellModel`` since that they impose material properties that are constant in time. However, it would affect models with properties that change in time, including aging viscoelastic models, or viscoelastic models combined with damage.

## Description

In ``GeneralizedKelvinVoingBase::updateQpViscousStrains`` the ``_springs_elasticity_tensors_inv`` need to be replaced with their old values (defined in the constructor but not stored).

In ``GeneralizedKelvinVoingBase::updateQpViscousStrains`` and ``GeneralizedMaxwellBase::updateQpViscousStrains`` the ``_driving_eigenstrain`` tensor needs to be replaced with its old value.

These old values need to be stored in ``LinearViscoelasticityBase``.

## Impact

Provides rigorous integration for aging or damaging viscoelastic materials.

No change of API required, but stores additional fields in ``LinearViscoelasticityBase``

### PR DESCRIPTION
**Pull request by agiorla**
- Update of viscous strains in linear viscoelastic materials now use the old tensors as they are supposed to be.
- For ``GeneralizedKelvinVoigtBase`` the old effective stress is directly back-calculated from the old elastic strain. This makes the scheme applicable even in the case of damage.
- The tests with driving eigenstrains have been regolded (there was a one-time-step delay in the previous version). The analytical expression of the strain expected in these tests have been provided to demonstrate the correctness of the calculation.

Closes #11064

### CODE REVIEW

File: modules/tensor_mechanics/src/materials/GeneralizedKelvinVoigtBase.C [hunk 1]
```diff
@@ -20,31 +20,38 @@ validParams<GeneralizedKelvinVoigtBase>()
...
+  : LinearViscoelasticityBase(parameters),
+    _first_elasticity_tensor_old(
+        getMaterialPropertyOld<RankFourTensor>(_base_name + "first_elasticity_tensor")),
+    _first_elasticity_tensor_inv_old(
+        getMaterialPropertyOld<RankFourTensor>(_base_name + "first_elasticity_tensor_inv")),
+    _springs_elasticity_tensors_inv_old(getMaterialPropertyOld<std::vector<RankFourTensor>>(
+        _base_name + "springs_elasticity_tensors_inv"))
 {
-  _need_viscoelastic_properties_inverse = true;
+  //  _need_viscoelastic_properties_inverse = true;
```
Comment (bwspenc) [inline modules/tensor_mechanics/src/materials/GeneralizedKelvinVoigtBase.C]: It looks like this gets set in the parameters, so there's no need to set it (which I assume is why you commented it out).  Delete this commented out line of code if that's the case.

File: modules/tensor_mechanics/test/tests/visco/gen_kv_driving.i [hunk 2]
```diff
@@ -1,3 +1,9 @@
+# represents a unique Maxwell module with E = 10GPa and eta = 10 days with an imposed eigenstrain alpha = 0.001
+# the specimen is free (sigma = 0) but the eigenstrain contributes to the creep deformation
```
Comment (bwspenc) [inline modules/tensor_mechanics/test/tests/visco/gen_kv_driving.i]: Is that really right?  It seems like the elastic strain would be zero, so you'd get no creep strain.

**Review by bwspenc — CHANGES_REQUESTED** [review]
Just a small change to make and a question.

File: modules/tensor_mechanics/src/materials/GeneralizedKelvinVoigtBase.C (hunk 1 above)
Comment (agiorla) [reply modules/tensor_mechanics/src/materials/GeneralizedKelvinVoigtBase.C]: Yes, it is set in the parameters (I commented it to check whether it was set correctly). I forgot to remove it...

File: modules/tensor_mechanics/test/tests/visco/gen_kv_driving.i (hunk 2 above)
Comment (agiorla) [reply modules/tensor_mechanics/test/tests/visco/gen_kv_driving.i]: This test (and its companion gen_maxwell_driving) represents materials in which the creep strain is driven by the stress + the equivalent stress imposed by the eigenstrain (the driving_eigenstrain variable). This technique has been used by some authors to represent drying creep in concrete (additional creep strain induced by drying shrinkage).

In this test, the elastic strain is indeed 0, but the driving eigenstrain is added to the elastic strain when performing the update of the viscous strains. So there is a creep strain (which is the same creep strain as you would have under an elastic strain equal to the driving eigenstrain), and the total strain (epsilon in the comment) is therefore different from the eigenstrain + the creep strain derived from the eigenstrain. I checked that in this condition, MOOSE indeed gives an elastic strain equal to 0 (1e-16 on my machine).

I can modify the comment to clarify that if needed.

Comment (bwspenc) [reply modules/tensor_mechanics/test/tests/visco/gen_kv_driving.i]: OK, that makes sense.  While you're making that other change, if you could explain that a little bit more in the test input file, that would be good. Other than that, I think this is good to go.

**Comment by agiorla** [comment]
@bwspenc I made the corrections. There are some BlackBear tests that are failing since now the integration in time is correct for aging materials, but they might just need regolding.

**Comment by bwspenc** [comment]
@agiorla Thanks for making those changes.  Could you please build BlackBear against your version of MOOSE and verify that the diffs are expected? If so, I'll go ahead and merge this, which will temporarily break BlackBear, and then please submit a PR to BlackBear to re-gold those tests.

**Comment by bwspenc** [comment]
I'm a little puzzled because the Grizzly tests pass just fine, and they are currently duplicates of the concrete tests in BlackBear.

**Comment by agiorla** [comment]
That is rather weird that the Grizzly tests pass...

I had a look into what is going on with BlackBear. I remembered that I already noticed this difference in BlackBear and tried to correct it in BlackBear, so now there is a kind of conflict between the two solutions.

In fact, here I have assumed that the viscosities of the different dashpots are constant in time, but in BlackBear we are in a case where they are not. **Please do not merge this PR**, I need to make sure it also works for time-dependent viscosities... I will commit only once I am convinced that it also works for Grizzly as intended.

**Comment by agiorla** [comment]
I've rewritten the scheme so that it does the right thing even when the viscosities are not constant in time (as it is the case for BlackBear). There will be a need to modify the ConcreteLogarithmicCreepModel in BlackBear but that is a rather simple fix (essentially, we need to evaluate the viscosities at the middle of the time step).

**Comment by agiorla** [comment]
I have the BlackBear changes ready. Should I start the BR for them as well?

**Comment by bwspenc** [comment]
@agiorla We'll merge this PR first, then submit the BlackBear changes. We'll also need to make the same changes to Grizzly since it's still both places (I need to fix that too, but let's just make the same fix both places for now).

**Review by bwspenc — APPROVED** [review]

**Comment by agiorla** [comment]
Sure! Just let me know when we can start the BlackBear PR. I imagine the App tests that are failing now are the Grizzly tests?

**Pull request merged**